### DISCUSSION TITLE
Folia Support

### DISCUSSION THREAD

**Discussion by MrButtersDEV** [GriefPrevention/GriefPrevention#2033] [Ideas]
I understand that Folia is new and very young in its stages but I would like to see Folia support for Grief Prevention. This is more of a long term discussion as I don't see support being added anytime soon but worth starting the discussion.

**Comment by Jikoo** [comment]
Folia is probably going to need to produce some kind of bridge for plugins to use before it's likely to happen - abstracting out all scheduler usage is a reasonably large undertaking that any plugin supporting Folia and non-Folia platforms will have to do, and creating that abstraction without intimate knowledge of how Folia actually works sounds like a recipe for disaster. That said, Robo also wasn't interested in adding a dependency on PaperLib despite its small size.

Folia does look super cool and I hope it revolutionizes large-scale servers, but I also am not personally willing to invest in figuring out adoption until it's had some time to stabilize.

**Reply by RoboMWM** [reply]
ya want to reduce dependencies. That being said, I think only scheduler usage is pvp-related checks? GP should have minimal scheduler usage otherwise as far as I'm aware. Oh right, claim expirations.

**Reply by iLemon** [reply]
We're planning on releasing a folia server in a couple weeks and believe it's production ready.

The most recent test which peaked 1000 players:
![image](https://github.com/TechFortress/GriefPrevention/assets/18317307/44e6ae77-200f-4f81-a7f8-967c5eb3669f)

Any interest in taking another look at supporting it? It's been a while since April and adding support seems more feasible now based on other open-source plugins :)

**Reply by sirlinium** [reply]
Hello, same question, would it be possible to envision supporting folia :) ?

**Reply by Navoei** [reply]
Recently (last update to the source code was three weeks ago), an easier way for developers to migrate to folia has appaeared. 
https://github.com/Anon8281/UniversalScheduler
However, I have not tested it yet.

**Reply by Jikoo** [reply]
I personally would use A248's [MorePaperLib](https://github.com/A248/MorePaperLib) instead - they have a history of being very helpful (and clearly very knowledgeable) in the realm of concurrency for plugins.

**Comment by Bobcat00** [comment]
> I think only scheduler usage is pvp-related checks? GP should have minimal scheduler usage otherwise as far as I'm aware. Oh right, claim expirations.

Delivering claim blocks is a scheduled task. Now what will happen if it's delivering claim blocks at the same time that someone is creating/resizing/deleting a claim? The number of claim blocks could be corrupted.

**Reply by Jikoo** [reply]
This is very true. GP in general is not thread-safe (a select few things like the claims list are), and while we could fix the immediate incompatibilities relatively easily with a scheduler library, without a rewrite of... pretty much everything, you'd be sitting on a time bomb. When that hits, if you're lucky you'd see console errors, and if you're unlucky affected claims would be corrupted on disk or possibly deleted.

**Reply by Bobcat00** [reply]
Ah, just throw a bunch of `synchronized` keywords in there. :-)
